File: core/src/main/java/ru/pocketbyte/locolaser/resource/file/BaseClassResourceFile.java
package ru.pocketbyte.locolaser.resource.file;

import ru.pocketbyte.locolaser.config.WritingConfig;
import ru.pocketbyte.locolaser.resource.PlatformResources;
import ru.pocketbyte.locolaser.resource.entity.*;
import ru.pocketbyte.locolaser.utils.TextUtils;

import java.io.File;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public abstract class BaseClassResourceFile extends ResourceStreamFile {

    protected abstract void writeHeaderComment(ResMap resMap, WritingConfig writingConfig) throws IOException ;
    protected abstract void writeClassHeader(ResMap resMap, WritingConfig writingConfig) throws IOException ;
    protected abstract void writeComment(WritingConfig writingConfig, String comment) throws IOException ;
    protected abstract void writeProperty(WritingConfig writingConfig, String propertyName, ResItem item) throws IOException ;
    protected abstract void writeClassFooter(ResMap resMap, WritingConfig writingConfig) throws IOException ;

    public BaseClassResourceFile(File file) {
        super(file);
    }

    @Override
    public ResMap read() {
        return null;
    }

    @Override
    public void write(ResMap resMap, WritingConfig writingConfig) throws IOException {
        ResLocale locale = resMap.get(PlatformResources.BASE_LOCALE);
        if (locale != null) {
            open();
            writeHeaderComment(resMap, writingConfig);
            writeln();
            writeClassHeader(resMap, writingConfig);

            Set<String> keysSet = new HashSet<>();
            for (ResItem item : locale.values()) {
                String propertyName = TextUtils.keyToProperty(item.key);
                if (!keysSet.contains(propertyName)) {

                    keysSet.add(propertyName);

                    ResValue valueOther = item.valueForQuantity(Quantity.OTHER);
                    if (valueOther != null && valueOther.value != null) {
                        writeComment(writingConfig, valueOther.value);
                    }

                    writeProperty(writingConfig, propertyName, item);
                }
            }

            writeClassFooter(resMap, writingConfig);
            close();
        }
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/resource/file/BaseClassResourceFile.kt
package ru.pocketbyte.locolaser.resource.file

import ru.pocketbyte.locolaser.config.ExtraParams
import ru.pocketbyte.locolaser.resource.Resources
import ru.pocketbyte.locolaser.resource.entity.*
import ru.pocketbyte.locolaser.resource.formatting.FormattingType
import ru.pocketbyte.locolaser.resource.formatting.NoFormattingType
import ru.pocketbyte.locolaser.utils.TextUtils

import java.io.File
import java.io.IOException
import java.util.HashSet
import java.util.regex.Pattern

abstract class BaseClassResourceFile(file: File) : ResourceStreamFile(file) {

    companion object {
        private val FORMATTING_PATTERN = Pattern.compile("%([0-9]*)([s|d])")
    }

    class FormatArgument(
        val index: Int?,
        val type: String
    )

    override val formattingType: FormattingType = NoFormattingType

    @Throws(IOException::class)
    protected abstract fun writeHeaderComment(resMap: ResMap, extraParams: ExtraParams?)

    @Throws(IOException::class)
    protected abstract fun writeClassHeader(resMap: ResMap, extraParams: ExtraParams?)

    @Throws(IOException::class)
    protected abstract fun writeComment(extraParams: ExtraParams?, comment: String)

    @Throws(IOException::class)
    protected abstract fun writeProperty(extraParams: ExtraParams?, propertyName: String, item: ResItem)

    @Throws(IOException::class)
    protected abstract fun writeClassFooter(resMap: ResMap, extraParams: ExtraParams?)

    override fun read(extraParams: ExtraParams?): ResMap? {
        return null
    }

    @Throws(IOException::class)
    override fun write(resMap: ResMap, extraParams: ExtraParams?) {
        val locale = resMap[Resources.BASE_LOCALE]
        if (locale != null) {
            open()
            writeHeaderComment(resMap, extraParams)
            writeln()
            writeClassHeader(resMap, extraParams)

            val keysSet = HashSet<String>()
            for (item in locale.values) {
                val propertyName = TextUtils.keyToProperty(item.key)
                if (!keysSet.contains(propertyName)) {

                    keysSet.add(propertyName)

                    val valueOther = item.valueForQuantity(Quantity.OTHER)
                    if (valueOther?.value != null) {
                        writeComment(extraParams, valueOther.value)
                    }

                    writeProperty(extraParams, propertyName, item)

                    var arguments = mutableListOf<Pair<Int, String>>()
                }
            }

            writeClassFooter(resMap, extraParams)
            close()
        }
    }
}
